src/service/history-impl.js

```diff
@@ -736,56 +834,86 @@ export class HistoryBindingVirtual_ {
   }
 
   /** @override */
-  cleanup_() {
+  cleanup() {
     this.unlistenOnHistoryPopped_();
   }
 
   /** @override */
-  setOnStackIndexUpdated(callback) {
-    this.onStackIndexUpdated_ = callback;
+  setOnStateUpdated(callback) {
+    this.onStateUpdated_ = callback;
   }
 
   /** @override */
-  push() {
-    // Current implementation doesn't wait for response from viewer.
-    this.updateStackIndex_(this.stackIndex_ + 1);
-    return this.viewer_.sendMessageAwaitResponse(
-        'pushHistory', dict({'stackIndex': this.stackIndex_})).then(() => {
-      return this.stackIndex_;
-    });
+  push(opt_stateUpdate) {
+    const message = /** @type {!JsonObject} */ (Object.assign({
+      'stackIndex': this.stackIndex_ + 1,
+    }, opt_stateUpdate || {}));
+    return this.viewer_.sendMessageAwaitResponse('pushHistory', message)
+        .then(response => {
+          this.updateHistoryState_(/** @type {!HistoryStateDef} */ (response));
+          return response;
```

Comment:
I'm not sure that the SRP viewer implements this response yet. We should have a fallback with current behavior in case.

Reply:
Done. I set it up now so it just uses the message (which at least includes the stackIndex and any changes to the state) if the response doesn't exist.